### PR TITLE
Added finetuning support for BERT based models on IMDB dataset.

### PR DESCRIPTION
**Pull request by quic-meetkuma**
- Added support for SequenceClassification task for BERT based models.
- Added support for IMDB dataset.

Note: It is under testing.

### CODE REVIEW

File: QEfficient/finetune/utils/train_utils.py [hunk 1]
```diff
@@ -175,18 +184,27 @@ def train(
                         ref_device="cpu",
                         ref_dtype=torch.float32,
                         # adjust atol & rtol this as required
-                        atol=1e-1,
+                        atol=1,
```
Comment (quic-meetkuma) [inline QEfficient/finetune/utils/train_utils.py:187]: change this value to 1e-1

**Review by quic-meetkuma — COMMENTED** [review]
Address minor comments along with the other reviewers' comments.

File: QEfficient/finetune/utils/train_utils.py (hunk 1 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Done

File: QEfficient/finetune/utils/config_utils.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
     PrefixTuningConfig,
 )
 from transformers import default_data_collator
-from transformers.data import DataCollatorForSeq2Seq
+from transformers.data import DataCollatorForSeq2Seq, DefaultDataCollator
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/config_utils.py]: Removed in latest commit.

File: QEfficient/finetune/utils/train_utils.py [hunk 3]
```diff
@@ -280,7 +301,21 @@ def train(
             else:
                 train_epoch_loss = total_loss / len(train_dataloader)
 
-        train_perplexity = torch.exp(train_epoch_loss)
+        if train_config.enable_ddp:
+            # Get the correct train loss from all the nodes.
+            dist.barrier()
+            dist.all_reduce(train_epoch_loss, op=dist.ReduceOp.SUM)
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/train_utils.py:307]: This will save the same recomputed value of train_epoch_loss and subsequently train_perplexity in all N metrics_data_*.json file for NxDDP. It is not a good idea to lose those individual values as it will not match with the train_step_loss values as per calculation.

File: QEfficient/finetune/utils/train_utils.py [hunk 4]
```diff
@@ -420,6 +466,83 @@ def evaluation(model, train_config, eval_dataloader, local_rank, tokenizer, devi
...
+
+    Args:
+        model: The model to evaluate
+        eval_dataloader: The dataloader containing the evaluation data
+        local_rank: The rank of the current node in a distributed setting
+        tokenizer: The tokenizer used to decode predictions
+
+    Returns: eval_acc, eval_epoch_loss
+    """
+    model.eval()
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/train_utils.py:481]: To maintain uniformity in the code, please use 'train_config.enable_ddp' flag in the if condition rather than 'local_rank'

File: QEfficient/finetune/utils/config_utils.py [hunk 5]
```diff
@@ -88,16 +87,14 @@ def get_dataloader_kwargs(train_config, dataset, dataset_processer, mode):
...
                 dataset, num_replicas=dist.get_world_size(), rank=dist.get_rank(), shuffle=True
             )
             kwargs["batch_size"] = batch_size
             kwargs["drop_last"] = True
-            kwargs["collate_fn"] = default_data_collator
     else:
         kwargs["batch_size"] = batch_size
         kwargs["drop_last"] = True
-        kwargs["collate_fn"] = default_data_collator
+    kwargs["collate_fn"] = DataCollatorForSeq2Seq(dataset_processer)
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/config_utils.py:99]: In #338 we have merged the changed for within batch padding, need to check if DataCollatorForSeq2Seq is used for same functionality, if yes then we should remove one to maintain the consistency of the code.

File: QEfficient/finetune/utils/train_utils.py (hunk 4 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py:481]: Updated. Thanks.

File: QEfficient/finetune/utils/train_utils.py [hunk 6]
```diff
@@ -103,6 +104,14 @@ def train(
     if train_config.enable_ddp:
         dist.broadcast(loss_0_counter, src=0)
 
+    acc_helper = None
+    if train_config.task_type == "seq_classification":
+        if local_rank is None:
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:109]: Please use train_config.enable_ddp flag here.

File: QEfficient/finetune/utils/train_utils.py [hunk 7]
```diff
@@ -359,7 +399,7 @@ def train(
     return results
 
 
-def evaluation(model, train_config, eval_dataloader, local_rank, tokenizer, device):
+def evaluation_ppl(model, train_config, eval_dataloader, local_rank, tokenizer, device):
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:402]: local_rank is unused in this function, please remove.

File: QEfficient/finetune/utils/train_utils.py [hunk 8]
```diff
@@ -420,6 +460,83 @@ def evaluation(model, train_config, eval_dataloader, local_rank, tokenizer, devi
     return eval_ppl, eval_epoch_loss, val_step_loss, val_step_perplexity
 
 
+def evaluation_acc(model, train_config, eval_dataloader, local_rank, tokenizer, device):
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:463]: local_rank is unused in this function, please remove.

Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:463]: this function has common code with evaluation_ppl,  need to merge both in single function and use flag task_type inside if else for acc/ppl computation.

**Review by quic-mamta — CHANGES_REQUESTED** [review]

File: QEfficient/finetune/utils/train_utils.py (hunk 6 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Done

File: QEfficient/finetune/utils/train_utils.py (hunk 8 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: removed

File: QEfficient/finetune/utils/train_utils.py (hunk 7 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: removed

File: QEfficient/finetune/utils/train_utils.py (hunk 8 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Yes, nice catch. Thanks. Updated in latest one.

**Review by quic-mamta — APPROVED** [review]

**Review by quic-swatia — APPROVED** [review]
